src/libawkward/array/NumpyArray.cpp

```diff
@@ -3619,4 +3982,143 @@ namespace awkward {
+      }
+      else if (ascending  &&  stable) {
+        std::stable_sort(start, stop,
+          [&data](size_t i1, size_t i2) {return data[i1] < data[i2];});
+      }
+      else if (!ascending  &&  stable) {
+        std::stable_sort(start, stop,
+          [&data](size_t i1, size_t i2) {return data[i1] > data[i2];});
+      }
+      std::transform(start, stop, start,
+                 [&](size_t j) -> size_t { return j - (size_t)offsets[i]; });
+    }
+
+    struct Error err2 = awkward_argsort_64(
+      ptr.get(),
+      &result[0],
+      length);
+    util::handle_error(err2, classname(), nullptr);
+
+    return ptr;
+  }
+
+  template<typename T>
+  const std::shared_ptr<void>
+  NumpyArray::array_sort(const T* data,
+                         int64_t length,
+                         int64_t offset,
+                         const Index64& starts,
+                         const Index64& parents,
+                         int64_t outlength,
+                         bool ascending,
+                         bool stable) const {
+    std::shared_ptr<T> ptr(
+      new T[(size_t)length], util::array_deleter<T>());
+
+    if (length == 0) {
+      return ptr;
+    }
+
+    std::vector<size_t> result(length);
+    std::iota(result.begin(), result.end(), 0);
+
+    std::vector<int64_t> offsets = sorting_ranges(parents,
+                                                  outlength);
+
+    for (int64_t i = 0; i < offsets.size() - 1; i++) {
+      auto start = std::next(result.begin(), offsets[i]);
+      auto stop = std::next(result.begin(), offsets[i + 1]);
+
+      if (ascending  &&  !stable) {
+        std::sort(start, stop,
+          [&data](size_t i1, size_t i2) {return data[i1] < data[i2];});
+      }
+      else if (!ascending  &&  !stable) {
+        std::sort(start, stop,
+          [&data](size_t i1, size_t i2) {return data[i1] > data[i2];});
+      }
+      else if (ascending  &&  stable) {
+        std::stable_sort(start, stop,
+          [&data](size_t i1, size_t i2) {return data[i1] < data[i2];});
+      }
+      else if (!ascending  &&  stable) {
+        std::stable_sort(start, stop,
+          [&data](size_t i1, size_t i2) {return data[i1] > data[i2];});
+      }
+    }
+
+    struct Error err2 = util::awkward_numpyarray_sort<T>(
+      ptr.get(),
+      data,
+      &result[0],
+      starts.ptr().get(),
+      parents.ptr().get(),
+      parents.offset(),
+      parents.length());
+    util::handle_error(err2, classname(), nullptr);
+
+    return ptr;
+  }
+
+  std::vector<int64_t>
+  NumpyArray::sorting_ranges(const Index64& parents,
+                             int64_t outlength) const {
+    std::vector<int64_t> result;
+    std::vector<int64_t> ranges(parents.length() + 1);
+    struct Error err = awkward_argsort_prepare_ranges(
+      &ranges[0],
+      parents.ptr().get(),
+      parents.offset(),
+      parents.length(),
+      outlength);
+    util::handle_error(err, classname(), nullptr);
```

Comment:
This is going to a kernel, and that's fine, but the interface should be an `Index64` instead of a `std::vector<int64_t>::data`, which would make more sense once the actual sorting goes onto the kernel side.

Reply:
done